### PR TITLE
bsdflags support: skip blk/chr/lnk file type, fixes #3130

### LINKED ISSUE #3130: 1.1.0: strange slowdown when backupping device node files
I have several Debian build-roots, all created with pbuilder/cowbuilder on one system.

During the backup of those build-roots, borgbackup slows down to a crawl, when it hits devices in <chroot>/dev/ where no hardware exists in the host, for example in my case the floppy drive devices.

strace shows this:

```
[...]
     0.000000 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1680", "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
     0.000370 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1680", "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
     0.000270 ioctl(1, TIOCGWINSZ, {ws_row=44, ws_col=166, ws_xpixel=1848, ws_ypixel=1030}) = 0
     0.000462 write(2, "15.85 GB O 8.82 GB C 245.00 MB D"..., 167) = 167
     0.000389 lstat("/var/cache/pbuilder/base.cow/dev/fd0u1722", {st_mode=S_IFBLK|0660, st_rdev=makedev(2, 60), ...}) = 0
     0.000218 open("/var/cache/pbuilder/base.cow/dev/fd0u1722", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC) = -1 ENXIO (No such device or address)
     3.082317 llistxattr("/var/cache/pbuilder/base.cow/dev/fd0u1722", "", 4096) = 0
     0.000133 open("/var/cache/pbuilder/base.cow/dev/fd0u1722", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC) = -1 ENXIO (No such device or address)
     3.315956 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1722", "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
     0.000138 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1722", "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
     0.000216 ioctl(1, TIOCGWINSZ, {ws_row=44, ws_col=166, ws_xpixel=1848, ws_ypixel=1030}) = 0
     0.000313 write(2, "15.85 GB O 8.82 GB C 245.00 MB D"..., 167) = 167
     0.000309 lstat("/var/cache/pbuilder/base.cow/dev/fd0u1743", {st_mode=S_IFBLK|0660, st_rdev=makedev(2, 76), ...}) = 0
     0.000175 open("/var/cache/pbuilder/base.cow/dev/fd0u1743", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC) = -1 ENXIO (No such device or address)
     3.085720 llistxattr("/var/cache/pbuilder/base.cow/dev/fd0u1743", "", 4096) = 0
     0.000083 open("/var/cache/pbuilder/base.cow/dev/fd0u1743", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC) = -1 ENXIO (No such device or address)
     3.316940 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1743", "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
     0.000113 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1743", "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
     0.000296 ioctl(1, TIOCGWINSZ, {ws_row=44, ws_col=166, ws_xpixel=1848, ws_ypixel=1030}) = 0
     0.000429 write(2, "15.85 GB O 8.82 GB C 245.00 MB D"..., 167) = 167
     0.000288 lstat("/var/cache/pbuilder/base.cow/dev/fd0u1760", {st_mode=S_IFBLK|0660, st_rdev=makedev(2, 96), ...}) = 0
     0.000167 open("/var/cache/pbuilder/base.cow/dev/fd0u1760", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC) = -1 ENXIO (No such device or address)
     3.075010 llistxattr("/var/cache/pbuilder/base.cow/dev/fd0u1760", "", 4096) = 0
     0.000162 open("/var/cache/pbuilder/base.cow/dev/fd0u1760", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC) = -1 ENXIO (No such device or address)
     3.063643 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1760", "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
     0.000138 getxattr("/var/cache/pbuilder/base.cow/dev/fd0u1760", "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
     0.000219 ioctl(1, TIOCGWINSZ, {ws_row=44, ws_col=166, ws_xpixel=1848, ws_ypixel=1030}) = 0
     0.000327 write(2, "15.85 GB O 8.82 GB C 245.00 MB D"..., 167) = 167
     0.000256 lstat("/var/cache/pbuilder/base.cow/dev/fd0u1840", {st_mode=S_IFBLK|0660, st_rdev=makedev(2, 116), ...}) = 0
     0.000165 open("/var/cache/pbuilder/base.cow/dev/fd0u1840", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC) = -1 ENXIO (No such device or address)
     3.083064 llistxattr("/var/cache/pbuilder/base.cow/dev/fd0u1840", "", 4096) = 0
[...]
```
Note the 3 second delay after each `open(".../dev/fd0...")`. (According to strace(1) the timestamp at the start of a line is the time elapsed since the start of the syscall on the line before.)

A normal "living" system wouldn't have any devices without the needed hardware, because `/dev` is a tmpfs populated by udev. But in the build-roots, borgbackup hits the real filesystem below the tmpfs with tons on unneeded devices nodes and things get strange.

In my case, the backup time regresses from ~45 seconds with 1.0.11 to over 12 minutes with 1.1.0.

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
opening a device file for a non-existing device can be very slow.

also: lstat -> stat(..., follow_symlinks=False) like everywhere else.

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
Not totally sure if "not (regularfile or directory)" is the best check.

**Comment by enkore** [comment]
Perhaps if BLK or CHR would be better.

**Comment by ThomasWaldmann** [comment]
My idea was to only deal with regular files and dirs (in a whitelist way) because I do not know what side effects there could be when opening these:
- symlink
- fifo
- socket
- door
- port
- and whatever else future comes up with

File: src/borg/platform/linux.pyx [hunk 1]
```diff
@@ -72,8 +72,11 @@ BSD_TO_LINUX_FLAGS = {
 
 
 def set_flags(path, bsd_flags, fd=None):
-    if fd is None and stat.S_ISLNK(os.lstat(path).st_mode):
```
Comment (ThomasWaldmann) [inline src/borg/platform/linux.pyx]: @enkore do you remember why links were skipped here?

the open() following has O_NOFOLLOW, so it should actually work on the symlink itself (if we have one).

Comment (ThomasWaldmann) [reply src/borg/platform/linux.pyx]: i tried on linux/ext4, lsattr/chattr not supported on a symlink.
the question now is whether that is fs specific or the case in general on linux.

Comment (enkore) [reply src/borg/platform/linux.pyx]: O_NOFOLLOW makes the open fail if it is attempting to open a symlink. The check here would simply avoid that error, but it is still required due to the race condition of the check.

Comment (ThomasWaldmann) [reply src/borg/platform/linux.pyx]: ok, i added the symlink check, to both get_flags and set_flags.

Comment (ThomasWaldmann) [reply src/borg/platform/linux.pyx]: ok, guess I am done here? collapse / merge?

**Pull request merged**